xarray/core/variable.py

```diff
@@ -404,34 +416,189 @@ def _item_key_to_tuple(self, key):
+                    variables.insert(i - len(slices), Variable((dim,), values))
+                else:
+                    slices.append((i, value))
+
+        try:
+            variables = _broadcast_compat_variables(*variables)
+        except ValueError:
+            raise IndexError("Dimensions of indexers mismatch: {}".format(key))
+
+        out_key = [variable.data for variable in variables]
+        out_dims = tuple(out_dims_set)
+        slice_positions = set()
+        for i, value in slices:
+            out_key.insert(i, value)
+            new_position = out_dims.index(self.dims[i])
+            slice_positions.add(new_position)
+
+        if slice_positions:
+            new_order = [i for i in range(len(out_dims))
+                         if i not in slice_positions]
+        else:
+            new_order = None
+
+        return out_dims, VectorizedIndexer(out_key), new_order
+
     def __getitem__(self, key):
         """Return a new Array object whose contents are consistent with
         getting the provided key from the underlying data.
 
-        NB. __getitem__ and __setitem__ implement "orthogonal indexing" like
-        netCDF4-python, where the key can only include integers, slices
-        (including `Ellipsis`) and 1d arrays, each of which are applied
-        orthogonally along their respective dimensions.
-
-        The difference does not matter in most cases unless you are using
-        numpy's "fancy indexing," which can otherwise result in data arrays
-        whose shapes is inconsistent (or just uninterpretable with) with the
-        variable's dimensions.
+        NB. __getitem__ and __setitem__ implement xarray-style indexing,
+        where if keys are unlabeled arrays, we index the array orthogonally
+        with them. If keys are labeled array (such as Variables), they are
+        broadcasted with our usual scheme and then the array is indexed with
+        the broadcasted key, like numpy's fancy indexing.
 
         If you really want to do indexing like `x[x > 0]`, manipulate the numpy
         array `x.values` directly.
         """
-        key = self._item_key_to_tuple(key)
-        key = indexing.expanded_indexer(key, self.ndim)
-        dims = tuple(dim for k, dim in zip(key, self.dims)
-                     if not isinstance(k, integer_types))
-        values = self._indexable_data[key]
-        # orthogonal indexing should ensure the dimensionality is consistent
-        if hasattr(values, 'ndim'):
-            assert values.ndim == len(dims), (values.ndim, len(dims))
-        else:
-            assert len(dims) == 0, len(dims)
-        return type(self)(dims, values, self._attrs, self._encoding,
+        dims, index_tuple, new_order = self._broadcast_indexes(key)
+        data = self._indexable_data[index_tuple]
+        if new_order:
+            data = np.moveaxis(data, range(len(new_order)), new_order)
+        assert getattr(data, 'ndim', 0) == len(dims), (data.ndim, len(dims))
```

Comment:
Can we raise a `ValueError` here instead and provide  more informative error message?

Reply:
Actually, I don't think of the case this assertion fails.
Maybe I will remove this line.